src/main/java/co/cask/spanner/source/SpannerSource.java

```diff
@@ -0,0 +1,259 @@
+
+  @Override
+  public void prepareRun(BatchSourceContext batchSourceContext) throws Exception {
+    config.validate();
+    Configuration configuration = new Configuration();
+    initializeConfig(configuration);
+
+    // initialize spanner
+    spanner = SpannerUtil.getSpannerService(config.serviceAccountFilePath, config.project);
+    BatchClient batchClient =
+      spanner.getBatchClient(DatabaseId.of(config.project, config.instance, config.database));
+    Timestamp logicalStartTimeMicros =
+      Timestamp.ofTimeMicroseconds(TimeUnit.MILLISECONDS.toMicros(batchSourceContext.getLogicalStartTime()));
+
+    // create batch transaction id
+    BatchReadOnlyTransaction batchReadOnlyTransaction =
+      batchClient.batchReadOnlyTransaction(TimestampBound.ofReadTimestamp(logicalStartTimeMicros));
+    BatchTransactionId batchTransactionId = batchReadOnlyTransaction.getBatchTransactionId();
+
+    // partitionQuery returns ImmutableList which doesn't implement java Serializable interface,
+    // we add to array list, which implements java Serializable
+    List<Partition> partitions = new ArrayList<>(batchReadOnlyTransaction.partitionQuery(getPartitionOptions(),
+                                                                                         Statement.of(config.query)));
+
+    // serialize batch transaction-id and partitions
+    configuration.set(SpannerConstants.SPANNER_BATCH_TRANSACTION_ID, getSerializedObjectString(batchTransactionId));
+    configuration.set(SpannerConstants.PARTITIONS_LIST, getSerializedObjectString(partitions));
+
+    // set input format and pass configuration
+    batchSourceContext.setInput(Input.of(config.referenceName,
+                                         new SourceInputFormatProvider(SpannerInputFormat.class, configuration)));
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    schema = config.getSchema();
+  }
+
+  @Override
+  public void transform(KeyValue<NullWritable, ResultSet> input, Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<Schema.Field> fields = schema.getFields();
+    ResultSet resultSet = input.getValue();
+    for (Schema.Field field : fields) {
+      String fieldName = field.getName();
+      Type columnType = resultSet.getColumnType(fieldName);
+      if (columnType == null || resultSet.isNull(fieldName)) {
+        continue;
+      }
+      switch (columnType.getCode()) {
+        case BOOL:
+          builder.set(fieldName, resultSet.getBoolean(fieldName));
+          break;
+        case INT64:
+          builder.set(fieldName, resultSet.getLong(fieldName));
+          break;
+        case FLOAT64:
+          builder.set(fieldName, resultSet.getDouble(fieldName));
+          break;
+        case STRING:
+          builder.set(fieldName, resultSet.getString(fieldName));
+          break;
+        case BYTES:
```

Comment:
We should handle date, datetime and timestamp

Reply:
we will add that in a separate PR.